### PR TITLE
Implemented critical hits and weapon specific crit chance system

### PR DESCRIPTION
**Pull request by R1ndT**
Hey everyone,
this PR should fix the issue #25. I am aware that others tried to implement the feature before me, but since their PR's are open for a few months now, so I have decided to take matter into my own hands.
Since I am fairly new to Java, I am sure that my code will be a bit chaotic and could certainly be executed better. If you happen to find some bugs or some obvious ways to improve the code, be sure to let me know.
I am aware that the crit system is quite overpowered. The best way to solve this issue - in my opinion - is implementing a weapon specific minimum and maximum crit damage multiplier (similar to the crit chance system I've implemented). The reason I did not implement this feature yet, is that it would add more attributes to all the weapons making the code a bit more chaotic. If you do not mind that, I will be more than happy to add it.

Best regards
RindT

### CODE REVIEW

File: src/com/hotmail/kalebmarc/textfighter/main/Enemy.java
Comment (hhaslam11) [inline src/com/hotmail/kalebmarc/textfighter/main/Enemy.java:27]: Make this a private var, and if you need to access it from another class create a `getter` for it :)

File: src/com/hotmail/kalebmarc/textfighter/main/Enemy.java
Comment (hhaslam11) [inline src/com/hotmail/kalebmarc/textfighter/main/Enemy.java:114]: although this works, its common to refer to non-static class variables with the `this` keyword. 
```suggestion
        this.damageDealt = damage;
```

File: src/com/hotmail/kalebmarc/textfighter/main/Enemy.java
Comment (hhaslam11) [inline src/com/hotmail/kalebmarc/textfighter/main/Enemy.java:113]: this naming is a bit confusing. "return" makes me think it's going to return a value; I would recommend changing the name, or adding a comment to explain it a little better. In this specific situation, I think it would make sense to name it `setDamageDealt`

File: src/com/hotmail/kalebmarc/textfighter/main/Weapon.java
Comment (hhaslam11) [inline src/com/hotmail/kalebmarc/textfighter/main/Weapon.java:246]: I would consider pulling this formula out into a separate `private function`, just to make it a little more clear to understand what's going on with this line.

something like this makes it immediately obvious what the code does
```java
if(wasCriticalHit()) {
```

**Review by hhaslam11 — CHANGES_REQUESTED** [review]
Looks good! There's some minor refactoring you can do, but overall it's great.

**Comment by hhaslam11** [comment]
> it would add more attributes to all the weapons making the code a bit more chaotic. If you do not mind that, I will be more than happy to add it.

Definitely feel free to go ahead and implement this!

File: src/com/hotmail/kalebmarc/textfighter/main/Enemy.java
Comment (R1ndT) [reply src/com/hotmail/kalebmarc/textfighter/main/Enemy.java:27]: Alright, I did not even know something like this existed until now :D I have to say that it is much more clever. Still, I posted a question in the commit resolving this issue. I would greatly appreciate if you answered it.

Comment (R1ndT) [reply src/com/hotmail/kalebmarc/textfighter/main/Enemy.java]: Also why can't I just make it a public variable and reference it anywhere? (This is probably common sense for you, but I do not really understand why not)

File: src/com/hotmail/kalebmarc/textfighter/main/Enemy.java
Comment (R1ndT) [reply src/com/hotmail/kalebmarc/textfighter/main/Enemy.java]: Does it have any particular reason, or does it make the code just look just a bit cleaner?

File: src/com/hotmail/kalebmarc/textfighter/main/Enemy.java
Comment (R1ndT) [reply src/com/hotmail/kalebmarc/textfighter/main/Enemy.java]: Yes, I am already aware of this issue at this moment. Still, thanks for pointing it out!

File: src/com/hotmail/kalebmarc/textfighter/main/Enemy.java
Comment (hhaslam11) [reply src/com/hotmail/kalebmarc/textfighter/main/Enemy.java]: > When getting the value, why do I have to type "Weapon.get().getDamageDealt()" instead of just "Weapon.getDamageDealt()", when the variable is not directly tied to the weapon?

Great question! This has to do with how Object-Oriented Programming works. Classes need to be initialized/created before you can use them. Think of them as templates. As I'm sure you're aware, you can have multiple instances of `Weapon`, each with its own values. The exception is `static` variables/methods, which is tied to the class itself, instead of the instance of the class. Since `getDamageDealt()` is not static, that means its tied to the instance and you can not access it directly (even though it changes with each round, it technically is still a property of `Weapon`).

`Weapon.get()` is a static method for getting the currently "held" weapon. When calling this, you are actually getting the instance of `Weapon`, and not `Weapon` itself

Let me know if this is still confusing, I can write a better example if needed!


> Also why can't I just make it a public variable and reference it anywhere? 
There's nothing "illegal" about doing it this way, it works just fine, a lot of projects will do it this way just because its easier. Though, it is common (and good) practice to not allow direct access to variables. There are plenty of benefits, including:
- Security. It can help prevent accidental changes to the variable (you might only want to allow reading the variable, not changing it)
- You can add validation if needed (using a setter to make sure code isn't trying to set it to a value it shouldn't be)
[here's a more technical answer](https://stackoverflow.com/a/1568230/3291305)

[check this out as well](https://www.w3schools.com/java/java_encapsulation.asp)

Note that most of these things usually aren't a problem in smaller projects like this, but become much more important in large projects with dozens of people sharing the same codebase; however, it's always smart to plan ahead and follow good coding styles!

Again, let me know if that still doesn't make sense and I can get better examples for you

File: src/com/hotmail/kalebmarc/textfighter/main/Enemy.java
Comment (hhaslam11) [reply src/com/hotmail/kalebmarc/textfighter/main/Enemy.java]: @R1ndT the only benefit, in this case, is it just makes it obvious that the variable is a class property, and I don't have to scroll up to see what it's coming from

In general, though, it can prevent accidentally accessing the wrong variable (maybe there's a variable with the same name in scope).

This is another case where it's not really a big deal in smaller projects, but as the project grows (or you work on very large projects), it is helpful to be very explicit about every action because the codebase can become confusing with size

File: src/com/hotmail/kalebmarc/textfighter/main/Enemy.java
Comment (hhaslam11) [reply src/com/hotmail/kalebmarc/textfighter/main/Enemy.java]: just wanted to note that this would be considered a `setter`! Just like `getter`, but.. well, you know

**Review by hhaslam11 — APPROVED** [review]
If this is ready, feel free to merge it in :) not sure if you want to add more to it or not, so I will leave it to you to merge

**Comment by R1ndT** [comment]
> 
> 
> If this is ready, feel free to merge it in :) not sure if you want to add more to it or not, so I will leave it to you to merge

I do not plan on adding anything else regarding this topic. From what I see, I can only close the pull request and not merge it, so if I am doing something wrong, let me know or feel free to merge the PR yourself.

**Comment by hhaslam11** [comment]
> > If this is ready, feel free to merge it in :) not sure if you want to add more to it or not, so I will leave it to you to merge
> 
> I do not plan on adding anything else regarding this topic. From what I see, I can only close the pull request and not merge it, so if I am doing something wrong, let me know or feel free to merge the PR yourself.
> I am adding some evidence here:
> image

Ah didn't realize, alright. Will merge :) Thanks!

File: src/com/hotmail/kalebmarc/textfighter/main/Enemy.java
Comment (hhaslam11) [reply src/com/hotmail/kalebmarc/textfighter/main/Enemy.java]: > Thank you so much for the answers, they have really helped me a lot.
No problem! I was in your situation a few years ago, so I definitely understand what it's like; I'm more than happy to help out where I can!

Feel free to reach out whenever, even with questions about your own projects (either code questions, or high-level concepts), or if you just want some code reviews on something! My email is available on my Github profile :)

**Pull request merged**
